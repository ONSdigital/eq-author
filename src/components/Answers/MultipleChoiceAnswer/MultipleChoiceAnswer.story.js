import React from "react";
import PropTypes from "prop-types";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { reject } from "lodash";
import MultipleChoiceAnswer from "./index";
import { CHECKBOX, RADIO } from "constants/answer-types";

const options = [
  {
    id: "0",
    label: "",
    value: "",
    description: ""
  },
  {
    id: "1",
    label: "",
    value: "",
    description: ""
  }
];

class MultipleChoiceAnswerWrapper extends React.Component {
  static propTypes = {
    type: PropTypes.string
  };

  state = {
    answer: {
      id: "0",
      options
    }
  };

  nextId = 1;

  handleAddOption = () => {
    const { answer } = this.state;
    const newOption = {
      ...options[0],
      id: (++this.nextId).toString()
    };

    const newState = {
      answer: {
        ...answer,
        options: answer.options.concat(newOption)
      }
    };

    this.setState(newState);
  };

  handleDeleteOption = id => {
    const { answer } = this.state;

    const newState = {
      answer: {
        ...answer,
        options: reject(answer.options, { id })
      }
    };

    this.setState(newState);
  };

  render() {
    return (
      <MultipleChoiceAnswer
        type={this.props.type}
        answer={this.state.answer}
        onUpdateOption={action("update option")}
        onUpdate={action("update")}
        onAddOption={this.handleAddOption}
        onDeleteOption={this.handleDeleteOption}
      />
    );
  }
}

storiesOf("AnswerTypes/MultipleChoiceAnswer", module)
  .add(RADIO, () => <MultipleChoiceAnswerWrapper type={RADIO} />)
  .add(CHECKBOX, () => <MultipleChoiceAnswerWrapper type={CHECKBOX} />);
