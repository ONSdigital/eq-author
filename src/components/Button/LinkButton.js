import React from "react";
import PropTypes from "prop-types";
import Button from "components/Button";
import { propTypes } from "components/Button/propTypes";

const StyledLink = Button.withComponent("a");

class LinkButton extends React.Component {
  static propTypes = {
    ...propTypes,
    href: PropTypes.string.isRequired
  };

  render() {
    const { children, href, ...otherProps } = this.props;
    return (
      <StyledLink
        {...otherProps}
        href={href}
        target="_blank"
        rel="noopener noreferrer"
      >
        {children}
      </StyledLink>
    );
  }
}

export default LinkButton;
