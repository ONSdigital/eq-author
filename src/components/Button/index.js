import React from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";
import { radius, colors } from "constants/theme";
import { darken } from "polished";

const darken10 = darken(0.1);

const ClearButton = css`
  background: transparent;
  border: none;
`;

const PrimaryButton = css`
  background-color: ${colors.blue};
  color: white;
  position: relative;
  border: none;
  &:focus,
  &:hover {
    background-color: ${darken10(colors.blue)}
  }
`;

const SecondaryButton = css`
  background-color: white;
  color: ${colors.text};
  border: 1px ${colors.borders} solid;
  &:focus,
  &:hover {
    background-color: ${colors.borders};
  }
`;

const TertiaryButton = css`
  background-color: ${colors.red};
  color: white;
  position: relative;
  border: none;

  &:focus,
  &:hover {
    background-color: ${darken10(colors.red)}
  }
`;

const StyledButton = styled.button`
  padding: 0.8em 3em;
  border-radius: ${radius};
  font-size: 0.8em;
  font-weight: 400;
  cursor: pointer;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  transition: background-color 200ms ease-out;
  letter-spacing: -0.2px;
  position: relative;
  overflow: hidden;

  &:focus,
  &:active {
    outline: none;
  }

  &[disabled] {
    pointer-events: none;
    opacity: 0.6;
  }

  ${props => props.primary && PrimaryButton} ${props =>
      props.secondary && SecondaryButton} ${props =>
      props.tertiary && TertiaryButton} ${props => props.clear && ClearButton};
`;

const Button = ({ children, type, ...otherProps }) =>
  <StyledButton {...otherProps} type={type}>
    {children}
  </StyledButton>;

Button.propTypes = {
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  tertiary: PropTypes.bool,
  type: PropTypes.oneOf(["button", "submit"]),
  clear: PropTypes.bool,
  children: PropTypes.node
};

export default Button;
