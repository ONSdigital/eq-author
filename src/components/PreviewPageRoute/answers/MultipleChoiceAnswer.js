import React from "react";
import styled, { css } from "styled-components";

import { CHECKBOX, RADIO } from "constants/answer-types";
import Error from "./ValidationError";
import { Field } from "./elements";

const Legend = styled.div`
  font-weight: bold;
  margin-bottom: 0.5em;
`;

const radioInput = css`
  border-radius: 100px;
  box-shadow: inset 0 0 0 3px #fff;
`;

const Input = styled.input`
  width: 20px;
  height: 20px;
  appearance: none;
  border: 2px solid #9b9b9b;
  padding: 0.5em;
  font-size: 1em;
  background: #eee;
  box-shadow: inset 0 0 0 3px white;
  pointer-events: none;
  position: absolute;
  top: 0.6em;
  left: 0.5em;

  ${props => props.type === RADIO && radioInput};
`;

const OptionLabel = styled.label`
  display: block;
  font-size: 1em;
  color: inherit;
  line-height: 1.4;
  font-weight: 400;
  padding: 0.7em 1em 0.7em 2.5em;
  margin: 0;
`;

const OptionDescription = styled.div`
  font-size: 0.8em;
  margin-top: 0.5em;
`;

const OptionItem = styled.div`
  font-size: 1em;
  background: #fff;
  border: 1px solid #999;
  border-radius: 0.2em;
  width: auto;
  min-width: 20em;
  display: inline-block;
  overflow: hidden;
  position: relative;
  margin-bottom: 0.25em;
`;

const SelectAll = styled.div`
  margin-bottom: 0.5em;
`;

const OtherLabel = styled.div`
  margin-bottom: 0.5em;
  font-size: 0.8em;
  font-weight: bold;
`;

const TextInput = styled.input`
  padding: 0.6em;
  display: block;
  color: inherit;
  font-size: 1em;
  border: 1px solid #999;
  border-radius: 3px;
  width: 100%;
`;

const OtherField = styled.div`
  padding: 0 0.5em 0.5em;
`;

const MutuallyExclusiveOption = styled.div`
  margin-top: 1em;
`;

const MutuallyExclusiveOptionTitle = styled.div`
  font-weight: bold;
  margin-bottom: 0.5em;
`;

const Option = ({ option, type, answer }) => (
  <OptionItem>
    <Input type={type} />
    <OptionLabel>
      {option.label || <Error>No label supplied</Error>}
      {option.description && (
        <OptionDescription>{option.description}</OptionDescription>
      )}
    </OptionLabel>
    {answer && (
      <OtherField>
        <OtherLabel>{answer.label}</OtherLabel>
        <TextInput type="text" />
      </OtherField>
    )}
  </OptionItem>
);

export default ({ answer }) => {
  return (
    <Field>
      <Legend>{answer.label}</Legend>
      {answer.type === CHECKBOX && (
        <SelectAll>Select all that apply:</SelectAll>
      )}
      {answer.options.map(option => (
        <Option key={option.id} option={option} type={answer.type} />
      ))}
      {answer.other && (
        <Option
          option={answer.other.option}
          answer={answer.other.answer}
          type={answer.type}
        />
      )}
      {answer.mutuallyExclusiveOption && (
        <MutuallyExclusiveOption>
          <MutuallyExclusiveOptionTitle>Or</MutuallyExclusiveOptionTitle>
          <Option option={answer.mutuallyExclusiveOption} type={answer.type} />
        </MutuallyExclusiveOption>
      )}
    </Field>
  );
};
