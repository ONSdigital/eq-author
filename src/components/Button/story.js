import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import styled from "styled-components";
import Button from "components/Button";

const Padding = styled.div`
  padding: 2em;
`;

storiesOf("Button", module)
  .addDecorator(story => <Padding>{story()}</Padding>)
  .add("Primary", () => (
    <Button onClick={action("click")} primary>
      Create survey
    </Button>
  ))
  .add("Secondary", () => (
    <Button onClick={action("click")} secondary>
      Cancel
    </Button>
  ))
  .add("Tertiary", () => (
    <Button onClick={action("click")} tertiary>
      Abort
    </Button>
  ));
