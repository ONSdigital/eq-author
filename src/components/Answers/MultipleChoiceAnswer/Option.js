import React, { Component } from "react";
import styled from "styled-components";
import { colors, radius } from "constants/theme";
import { Field } from "components/Forms";
import WrappingInput from "components/WrappingInput";
import withEntityEditor from "components/withEntityEditor";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import DeleteButton from "components/DeleteButton";
import Tooltip from "components/Tooltip";
import { CHECKBOX, RADIO } from "constants/answer-types";
import { get } from "lodash";
import optionFragment from "graphql/fragments/option.graphql";
import getIdForObject from "utils/getIdForObject";

const ENTER_KEY = 13;

export const DeleteContainer = styled.div`
  padding: 0.2em;
  position: absolute;
  top: 0.5em;
  right: 1em;
`;

const borderRadii = {
  [CHECKBOX]: "3px",
  [RADIO]: "100%"
};

const DummyInput = styled.div`
  border: ${radius} solid ${colors.borders};
  height: 1.4em;
  width: 1.4em;
  display: inline-block;
  margin: 0 1em 0 0;
  vertical-align: top;
  border-radius: ${props => get(borderRadii, props.type, "initial")};
`;

export const StyledOption = styled.div`
  border: 1px solid ${colors.borders};
  padding: 1em 1.8em 0 1em;
  border-radius: 3px;
  position: relative;

  &:not(:first-child) {
    margin-top: 0.5em;
  }

  &.option-enter {
    opacity: 0;
    height: 0;
    transform: translateX(-20px);
  }

  &.option-enter-active {
    transition: height ${props => props.duration / 2}ms ease-out,
      opacity ${props => props.duration / 2}ms ease-out
        ${props => props.duration / 2}ms,
      transform ${props => props.duration / 2}ms ease-out
        ${props => props.duration / 2}ms;
    opacity: 1;
    height: 5.625em;
    transform: translateX(0);
  }

  &.option-exit {
    opacity: 1;
    height: 5.625em;
    transform: translateX(0);
  }

  &.option-exit-active {
    transition: opacity ${props => props.duration / 2}ms ease-out,
      transform ${props => props.duration / 2}ms ease-out,
      height ${props => props.duration / 2}ms ease-out
        ${props => props.duration / 2}ms;
    opacity: 0;
    height: 0;
    transform: translateX(-20px);
  }
`;

StyledOption.defaultProps = {
  duration: 200
};

StyledOption.propTypes = {
  duration: PropTypes.number
};

export const SeamlessLabel = styled(WrappingInput)`
  display: inline-block !important;
  width: auto;
  vertical-align: middle;
  margin-right: 1.5em;
  flex: 1 1 0%;
`;

const LabelField = styled(Field)`
  display: flex;
`;

export class StatelessOption extends Component {
  static propTypes = {
    option: CustomPropTypes.option.isRequired,
    onChange: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    onEnterKey: PropTypes.func,
    hasDeleteButton: PropTypes.bool.isRequired,
    type: PropTypes.oneOf([RADIO, CHECKBOX]).isRequired,
    children: PropTypes.node,
    labelPlaceholder: PropTypes.string,
    autoFocus: PropTypes.bool
  };

  static defaultProps = {
    labelPlaceholder: "Label",
    autoFocus: true
  };

  handleDeleteClick = e => {
    this.props.onDelete(this.props.option.id);
  };

  handleKeyDown = e => {
    if (e.keyCode === ENTER_KEY) {
      this.props.onEnterKey(e);
    }
  };

  renderDeleteButton() {
    return (
      <DeleteContainer>
        <Tooltip content="Delete option">
          <DeleteButton
            size="small"
            aria-label="Delete option"
            onClick={this.handleDeleteClick}
            data-test="btn-delete-option"
          />
        </Tooltip>
      </DeleteContainer>
    );
  }

  render() {
    const {
      hasDeleteButton,
      option,
      onChange,
      onUpdate,
      type,
      children,
      labelPlaceholder,
      autoFocus
    } = this.props;

    return (
      <StyledOption id={getIdForObject(option)} key={option.id}>
        <LabelField>
          <DummyInput type={type} />
          <SeamlessLabel
            name="label"
            placeholder={labelPlaceholder}
            size="medium"
            value={option.label}
            onChange={onChange}
            onBlur={onUpdate}
            onKeyDown={this.handleKeyDown}
            data-test="option-label"
            data-autofocus={autoFocus || null}
          />
        </LabelField>
        <Field>
          <WrappingInput
            name="description"
            placeholder="Optional description"
            onChange={onChange}
            value={option.description}
            onBlur={onUpdate}
            onKeyDown={this.handleKeyDown}
            data-test="option-description"
          />
        </Field>
        {children}
        {hasDeleteButton && this.renderDeleteButton()}
      </StyledOption>
    );
  }
}

export default withEntityEditor("option", optionFragment)(StatelessOption);
