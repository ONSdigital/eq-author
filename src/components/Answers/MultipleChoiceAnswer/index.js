import React, { Component } from "react";

import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import { TransitionGroup } from "react-transition-group";
import styled from "styled-components";
import focusOnEntity from "utils/focusOnEntity";

import Option from "./Option";
import OptionTransition from "./OptionTransition";
import BasicAnswer from "components/Answers/BasicAnswer";

import { colors } from "constants/theme";
import { CHECKBOX } from "constants/answer-types";
import SplitButton from "components/SplitButton";
import Dropdown from "components/SplitButton/Dropdown";
import MenuItem from "components/SplitButton/MenuItem";

import { last } from "lodash";
import gql from "graphql-tag";

const AnswerWrapper = styled.div`
  margin: 2em 0 0;
  width: 75%;
`;

const AnswerHelper = styled.div`
  margin-bottom: 0.5em;
  font-size: 0.9em;
  font-weight: 600;
  color: ${colors.text};
`;

const Options = styled.div`
  margin: 0 0 1em;
`;

export const AddOtherLink = styled.button`
  color: #48a6f6;
  text-decoration: none;
  border: 0;
  background: none;
`;

const OtherAnswerWrapper = styled.div`
  padding-top: 0.25em;
  margin-bottom: 1em;
`;

class MultipleChoiceAnswer extends Component {
  static propTypes = {
    answer: CustomPropTypes.answer.isRequired,
    onUpdate: PropTypes.func.isRequired,
    onAddOption: PropTypes.func.isRequired,
    onUpdateOption: PropTypes.func.isRequired,
    onDeleteOption: PropTypes.func.isRequired,
    onAddOther: PropTypes.func.isRequired,
    onDeleteOther: PropTypes.func.isRequired,
    minOptions: PropTypes.number.isRequired
  };

  static defaultProps = {
    minOptions: 1
  };

  state = {
    open: false
  };

  handleOptionDelete = optionId => {
    this.props.onDeleteOption(optionId, this.props.answer.id);
  };

  handleAddOption = e => {
    e.preventDefault();
    e.stopPropagation();
    return this.props.onAddOption(this.props.answer.id).then(focusOnEntity);
  };

  handleAddOther = e => {
    e.preventDefault();
    return this.props
      .onAddOther(this.props.answer)
      .then(res => {
        this.handleToggleOpen(false);
        return res;
      })
      .then(res => res.option)
      .then(focusOnEntity);
  };

  handleDeleteOther = e => {
    return this.props
      .onDeleteOther(this.props.answer)
      .then(() => last(this.props.answer.options))
      .then(focusOnEntity);
  };

  handleToggleOpen = open => {
    this.setState({
      open
    });
  };

  render() {
    const {
      answer,
      onUpdateOption,
      onUpdate,
      minOptions,
      ...otherProps
    } = this.props;

    const numberOfOptions = answer.options.length + (answer.other ? 1 : 0);
    const showDeleteOption = numberOfOptions > minOptions;

    return (
      <BasicAnswer
        answer={answer}
        onUpdate={onUpdate}
        autoFocus={false}
        labelText="Label (optional)"
      >
        <AnswerWrapper>
          {answer.type === CHECKBOX && (
            <AnswerHelper>Select all that apply</AnswerHelper>
          )}
          <TransitionGroup component={Options}>
            {answer.options.map((option, optionIndex, options) => (
              <OptionTransition key={option.id}>
                <Option
                  {...otherProps}
                  option={option}
                  onDelete={this.handleOptionDelete}
                  onUpdate={onUpdateOption}
                  onEnterKey={this.handleAddOption}
                  hasDeleteButton={showDeleteOption}
                />
              </OptionTransition>
            ))}
            {answer.other && (
              <OptionTransition key={answer.other.option.id}>
                <Option
                  {...otherProps}
                  option={answer.other.option}
                  onDelete={this.handleDeleteOther}
                  onUpdate={onUpdateOption}
                  onEnterKey={this.handleAddOption}
                  hasDeleteButton={showDeleteOption}
                  labelPlaceholder="eg. Other"
                >
                  <OtherAnswerWrapper data-test="other-answer">
                    <BasicAnswer
                      answer={answer.other.answer}
                      onUpdate={onUpdate}
                      showDescription={false}
                      labelText="Other label"
                      labelPlaceholder="eg. Please specify"
                      bold={false}
                    />
                  </OtherAnswerWrapper>
                </Option>
              </OptionTransition>
            )}
          </TransitionGroup>
          <div>
            <SplitButton
              onPrimaryAction={this.handleAddOption}
              primaryText="Add another option"
              onToggleOpen={this.handleToggleOpen}
              open={this.state.open}
              dataTest="btn-add-option"
            >
              <Dropdown>
                <MenuItem
                  onClick={this.handleAddOther}
                  disabled={answer.other !== null}
                  data-test="btn-add-option-other"
                >
                  Add &quot;other&quot; option
                </MenuItem>
              </Dropdown>
            </SplitButton>
          </div>
        </AnswerWrapper>
      </BasicAnswer>
    );
  }
}

MultipleChoiceAnswer.fragments = {
  MultipleChoice: gql`
    fragment MultipleChoice on Answer {
      ...Answer
      ... on MultipleChoiceAnswer {
        options {
          ...Option
        }
        other {
          option {
            ...Option
          }
          answer {
            ...Answer
          }
        }
      }
    }
    ${Option.fragments.Option}
    ${BasicAnswer.fragments.Answer}
  `
};

export default MultipleChoiceAnswer;
