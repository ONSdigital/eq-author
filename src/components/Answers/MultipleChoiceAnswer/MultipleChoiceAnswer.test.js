import React from "react";

import Button from "components/Button";
import MultipleChoiceAnswer from "./";
import Option from "./Option";
import { times } from "lodash";

import { shallow } from "enzyme";

describe("MultipleChoiceAnswer", () => {
  let wrapper;

  let answer = {
    id: 0,
    options: [
      {
        id: 1,
        label: "",
        description: ""
      }
    ]
  };

  let mockHandlers = {
    onAddOption: jest.fn(),
    onUpdateOption: jest.fn(),
    onDeleteOption: jest.fn(),
    onChange: jest.fn()
  };

  const createWrapper = ({ answer }, render = shallow) =>
    render(<MultipleChoiceAnswer {...mockHandlers} answer={answer} />);

  beforeEach(() => {
    wrapper = createWrapper({ answer });
  });

  it("should have one option by default", () => {
    expect(wrapper.find(Option)).toHaveLength(1);
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should add a new option when add button is clicked", () => {
    const preventDefault = jest.fn();
    wrapper.find(Button).first().simulate("click", { preventDefault });
    expect(mockHandlers.onAddOption).toHaveBeenCalledWith(answer.id);
  });

  describe("delete button", () => {
    const minOptions = 2;

    const createAnswer = numberOptions => ({
      ...answer,
      options: times(numberOptions, id => ({
        id: id,
        label: "",
        description: ""
      }))
    });

    beforeEach(() => {
      wrapper.setProps({ minOptions });
    });

    it("should not show when number options <= minOptions", () => {
      const answer = createAnswer(minOptions);
      wrapper.setProps({ answer });

      wrapper.find(Option).forEach(option => {
        expect(option.prop("hasDeleteButton")).toBe(false);
      });
    });

    it("should show when number options > minOptions", () => {
      const answer = createAnswer(minOptions + 1);
      wrapper.setProps({ answer });

      wrapper.find(Option).forEach(option => {
        expect(option.prop("hasDeleteButton")).toBe(true);
      });
    });

    it("should handle deleting an option", () => {
      const optionId = answer.options[0].id;
      wrapper.find(Option).first().simulate("delete", optionId);

      expect(mockHandlers.onDeleteOption).toHaveBeenCalledWith(
        optionId,
        answer.id
      );
    });
  });
});
