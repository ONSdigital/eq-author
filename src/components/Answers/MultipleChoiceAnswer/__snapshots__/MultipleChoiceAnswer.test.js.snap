// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleChoiceAnswer should match snapshot 1`] = `
<withEntityEditor(StatelessBasicAnswer)
  answer={
    Object {
      "id": "0",
      "options": Array [
        Object {
          "__typename": "Option",
          "description": "",
          "id": "1",
          "label": "",
        },
      ],
    }
  }
  labelPlaceholder="Label (optional)"
  onChange={[Function]}
  onUpdate={[Function]}
>
  <MultipleChoiceAnswer__AnswerWrapper>
    <TransitionGroup
      childFactory={[Function]}
      component={[Function]}
    >
      <CSSTransition
        classNames="option"
        key="1"
        timeout={200}
      >
        <withEntityEditor(StatelessOption)
          hasDeleteButton={false}
          onAddOption={[Function]}
          onDelete={[Function]}
          onDeleteOption={[Function]}
          onEnterKey={[Function]}
          onUpdate={[Function]}
          option={
            Object {
              "__typename": "Option",
              "description": "",
              "id": "1",
              "label": "",
            }
          }
        />
      </CSSTransition>
    </TransitionGroup>
    <div>
      <Button
        data-test="btn-add-option"
        onClick={[Function]}
        secondary={true}
        type="button"
      >
        Add another option
      </Button>
    </div>
  </MultipleChoiceAnswer__AnswerWrapper>
</withEntityEditor(StatelessBasicAnswer)>
`;
